File: mifosng-android/src/instrumentTest/java/com/mifos/mifosxdroid/tests/CreateNewGroupFragmentTest.java
package com.mifos.mifosxdroid.tests;

import android.view.KeyEvent;
import android.widget.Spinner;

import androidx.test.core.app.ActivityScenario;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;
import androidx.test.rule.ActivityTestRule;

import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.online.DashboardActivity;
import com.mifos.mifosxdroid.online.createnewgroup.CreateNewGroupFragment;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.text.SimpleDateFormat;
import java.util.Date;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.hasMinimumChildCount;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
import static junit.framework.TestCase.assertNotNull;
import static junit.framework.TestCase.assertTrue;

@RunWith(AndroidJUnit4ClassRunner.class)
public class CreateNewGroupFragmentTest {

    private ActivityScenario activityScenario;
    @Rule
    public ActivityTestRule<DashboardActivity> activityTestRule =
            new ActivityTestRule<>(DashboardActivity.class);

    @Before
    public void setUp() throws Exception {
        activityScenario = ActivityScenario.launch(DashboardActivity.class);
        openCreateGroup();
    }

    @Test
    public void testViewsAreVisible() throws InterruptedException {

        Thread.sleep(3000);

        onView(withId(R.id.et_group_name))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.sp_group_offices))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.tv_group_submission_date))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.et_group_external_id))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.cb_group_active_status))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
        onView(withId(R.id.btn_submit))
                .perform(scrollTo())
                .check(matches(isDisplayed()));

        onView(withId(R.id.cb_group_active_status))
                .perform(scrollTo(), click())
                .check(matches(isDisplayed()));
        onView(withId(R.id.tv_group_activationDate))
                .perform(scrollTo())
                .check(matches(isDisplayed()));
    }

    @Test
    public void testViewsAreNotNull() throws InterruptedException {
        assertNotNull(onView(withId(R.id.et_group_name)));
        assertNotNull(onView(withId(R.id.sp_group_offices)));
        assertNotNull(onView(withId(R.id.tv_group_submission_date)));
        assertNotNull(onView(withId(R.id.et_group_external_id)));
        assertNotNull(onView(withId(R.id.cb_group_active_status)));
        assertNotNull(onView(withId(R.id.btn_submit)));
        assertNotNull(onView(withId(R.id.tv_group_activationDate)));
    }

    @Test
    public void testOpenCreateGroup() throws InterruptedException {

        CreateNewGroupFragment createNewGroupFragment
                = (CreateNewGroupFragment) activityTestRule.getActivity()
                .getSupportFragmentManager().findFragmentById(R.id.container);

        assertNotNull(createNewGroupFragment);
        assertTrue(createNewGroupFragment.isAdded());

    }

    @Test
    public void testSpinnerPopulated() throws InterruptedException {
        Thread.sleep(3000);
        onView(withId(R.id.sp_group_offices)).
                check(matches(hasMinimumChildCount(1)));
    }

    @Test
    public void testCreateGroup() throws InterruptedException {
        Thread.sleep(3000);

        // Group Name
        onView(withId(R.id.et_group_name))
                .perform(scrollTo())
                .perform(typeText("myTestGroup"), closeSoftKeyboard());

        // Office
        Spinner spinnerOffice = activityTestRule.getActivity().findViewById(R.id.sp_group_offices);
        if (spinnerOffice.getAdapter().getCount() > 2) {
            String selectedOffice = (String) spinnerOffice.getAdapter().getItem(2);
            onView(withId(R.id.sp_group_offices))
                    .perform(scrollTo(), click());
            onView(withText(selectedOffice))
                    .check(matches(isDisplayed()))
                    .perform(click())
                    .check(matches(isDisplayed()));
        }

        // Submission Date
        SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
        String currentDate = formatter.format(new Date());
        onView(withId(R.id.tv_group_submission_date))
                .perform(scrollTo())
                .check(matches(withText(currentDate)));

        // External Id
        onView(withId(R.id.et_group_external_id))
                .perform(scrollTo())
                .perform(typeText("123"), closeSoftKeyboard());

        // Activation Date
        onView(withId(R.id.cb_group_active_status))
                .perform(scrollTo(), click());
        onView(withId(R.id.tv_group_activationDate))
                .check(matches(withText(currentDate)));

        // Submit
        onView(withId(R.id.btn_submit))
                .perform(scrollTo(), click());
    }

    private void openCreateGroup() throws InterruptedException {
        getInstrumentation().sendKeyDownUpSync(KeyEvent.KEYCODE_MENU);
        onView(withText("Create Group"))
                .perform(click());
    }
}

File: mifosng-android/src/instrumentTest/java/com/mifos/mifosxdroid/tests/CreateNewGroupFragmentTest.kt
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.mifosxdroid.tests

import android.test.ActivityInstrumentationTestCase2
import android.view.KeyEvent
import android.view.View
import com.mifos.mifosxdroid.R
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher
import java.text.SimpleDateFormat
import java.util.Date

/**
 * Created by ahmed fathy on 20/04/16.
 */
class CreateNewGroupFragmentTest : ActivityInstrumentationTestCase2<DashboardActivity?>(
    DashboardActivity::class.java
) {
    private var mActivity: DashboardActivity? = null
    @Throws(Exception::class)
    fun setUp() {
        super.setUp()
        mActivity = getActivity()
        openCreateGroup()
    }

    /**
     * - chooses "Create Group" from options menu
     * - checks the CrateNewGroupFragment is added
     */
    @SmallTest
    @Throws(InterruptedException::class)
    fun testOpenCreateGroup() {
        val createNewGroupFragment: CreateNewGroupFragment = mActivity.getSupportFragmentManager()
            .findFragmentById(R.id.container) as CreateNewGroupFragment
        assertNotNull(createNewGroupFragment)
        assertTrue(createNewGroupFragment.isAdded())
    }

    @SmallTest
    @Throws(InterruptedException::class)
    fun testViewsVisible() {

        // initially, the group name, office spinner. submission date,
        // external id, active checkbox and submit button are visible
        Espresso.onView(ViewMatchers.withId(R.id.et_group_name))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(ViewMatchers.withId(R.id.sp_group_offices))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(ViewMatchers.withId(R.id.tv_group_submission_date))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(ViewMatchers.withId(R.id.et_group_external_id))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(ViewMatchers.withId(R.id.cb_group_active_status))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(ViewMatchers.withId(R.id.btn_submit))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        // initially, the activation date is invisible but shows after checking the active checkbox
        Espresso.onView(ViewMatchers.withId(R.id.cb_group_active_status))
            .perform(ViewActions.scrollTo(), ViewActions.click())
            .check(ViewAssertions.matches(ViewMatchers.isChecked()))
        Espresso.onView(ViewMatchers.withId(R.id.tv_group_activationDate))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    /**
     * checks that the office spinner is loaded with data from the template
     */
    @SmallTest
    @Throws(InterruptedException::class)
    fun testSpinnerPopulated() {
        Thread.sleep(3000)
        Espresso.onView(ViewMatchers.withId(R.id.sp_group_offices))
            .check(ViewAssertions.matches(hasChildren()))
    }

    /**
     * fills the template and presses submit
     */
    @MediumTest
    @Throws(InterruptedException::class)
    fun testCreateGroup() {
        Thread.sleep(3000)

        // group name
        Espresso.onView(ViewMatchers.withId(R.id.et_group_name))
            .perform(ViewActions.scrollTo())
            .perform(ViewActions.typeText("myTestGroup"), ViewActions.closeSoftKeyboard())

        // office
        val spinnerOffice: Spinner = mActivity.findViewById(R.id.sp_group_offices) as Spinner
        if (spinnerOffice.getAdapter().getCount() > 2) {
            val selectedOffice = spinnerOffice.getAdapter().getItem(2) as String
            Espresso.onView(ViewMatchers.withId(R.id.sp_group_offices))
                .perform(ViewActions.scrollTo(), ViewActions.click())
            Espresso.onView(ViewMatchers.withText(selectedOffice))
                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
                .perform(ViewActions.click())
                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        }

        // submission date
        val formatter = SimpleDateFormat("dd-MM-yyyy")
        val currentDate = formatter.format(Date())
        Espresso.onView(ViewMatchers.withId(R.id.tv_group_submission_date))
            .perform(ViewActions.scrollTo())
            .check(ViewAssertions.matches(ViewMatchers.withText(currentDate)))

        // external id
        Espresso.onView(ViewMatchers.withId(R.id.et_group_external_id))
            .perform(ViewActions.scrollTo())
            .perform(ViewActions.typeText("123"), ViewActions.closeSoftKeyboard())

        // activation date
        Espresso.onView(ViewMatchers.withId(R.id.cb_group_active_status))
            .perform(ViewActions.scrollTo(), ViewActions.click())
        Espresso.onView(ViewMatchers.withId(R.id.tv_group_activationDate))
            .check(ViewAssertions.matches(ViewMatchers.withText(currentDate)))

        // submit
        Espresso.onView(ViewMatchers.withId(R.id.btn_submit))
            .perform(ViewActions.scrollTo(), ViewActions.click())
    }

    /**
     * chooses "Create Group" from the options menu
     */
    @Throws(InterruptedException::class)
    private fun openCreateGroup() {
        getInstrumentation().sendKeyDownUpSync(KeyEvent.KEYCODE_MENU)
        Espresso.onView(ViewMatchers.withText("Create Group"))
            .perform(ViewActions.click())
    }

    companion object {
        /**
         * checks that the spinner is loaded with 1 or more items
         */
        fun hasChildren(): Matcher<View> {
            return object : TypeSafeMatcher<View>() {
                public override fun matchesSafely(view: View): Boolean {
                    return (view as Spinner).getAdapter().getCount() > 0
                }

                override fun describeTo(description: Description) {
                    description.appendText("The spinner has no children")
                }
            }
        }
    }
}